Zend/tests/match/013.phpt

```diff
@@ -0,0 +1,93 @@
+--TEST--
+Match expression string jump table
+--INI--
+opcache.enable=1
+opcache.enable_cli=1
+opcache.opt_debug_level=0x20000
+--SKIPIF--
+<?php if (!extension_loaded('Zend OPcache')) die('skip Zend OPcache extension not available'); ?>
+--FILE--
+<?php
+
+function test($char) {
+    return match ($char) {
+        'a' => 'a',
+        'b', 'c' => 'b, c',
+        'd' => 'd',
+        'e', 'f' => 'e, f',
+        'g' => 'g',
+        'h', 'i' => 'h, i',
+    };
+}
+
+foreach (range('a', 'i') as $char) {
+    var_dump(test($char));
+}
+
+--EXPECTF--
+$_main:
+     ; (lines=%d, args=0, vars=1, tmps=2)
+     ; (after optimizer)
+     ; %s
+0000 INIT_FCALL 2 112 string("range")
+0001 SEND_VAL string("a") 1
+0002 SEND_VAL string("i") 2
+0003 V2 = DO_ICALL
+0004 V1 = FE_RESET_R V2 0013
+0005 FE_FETCH_R V1 CV0($char) 0013
+0006 INIT_FCALL 1 96 string("var_dump")
+0007 INIT_FCALL 1 112 string("test")
+0008 SEND_VAR CV0($char) 1
+0009 V2 = DO_UCALL
+0010 SEND_VAR V2 1
+0011 DO_ICALL
+0012 JMP 0005
+0013 FE_FREE V1
+0014 RETURN int(1)
+LIVE RANGES:
+     1: 0005 - 0013 (loop)
+
+test:
+     ; (lines=%d, args=1, vars=1, tmps=1)
+     ; (after optimizer)
+     ; %s
+0000 CV0($char) = RECV 1
+0001 SWITCH_STRING CV0($char) "a": 0020, "b": 0021, "c": 0021, "d": 0022, "e": 0023, "f": 0023, "g": 0024, "h": 0025, "i": 0025, default: 0026
```

Comment:
The generated opcodes look good to me but share an inefficiency with `switch` that **may or may not be worth improving in a followup PR if this gets merged:**
if the value passed to `SWITCH_STRING` **isn't a string, it will go to the next opcode** (ZEND_VM_NEXT_OPCODE) (the next opcode of `0001` is `0002`, then for something that isn't even a string, the VM will call IS_IDENTICAL '"a", then IS_IDENTICAL "b", etc. -- all of the comparisons would be performed and fail before falling through to the default or throw UnhandledMatchError case.

- Same for SWITCH_LONG.
- For switch, it needs to do that because it used `==` (loose equality), but this checks if values are identical.

See `Zend/zend_vm_def.h` for details - If you decide to edit it Zend/zend_vm_gen.php would be run to regenerate other files.

```
ZEND_VM_COLD_CONSTCONST_HANDLER(188, ZEND_SWITCH_STRING, CONST|TMPVARCV, CONST, JMP_ADDR)
{
	USE_OPLINE
	zval *op, *jump_zv;
	HashTable *jumptable;

	op = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);
	jumptable = Z_ARRVAL_P(GET_OP2_ZVAL_PTR(BP_VAR_R));

	if (Z_TYPE_P(op) != IS_STRING) {
		if (OP1_TYPE == IS_CONST) {
			/* Wrong type, fall back to ZEND_CASE chain */
			ZEND_VM_NEXT_OPCODE();
		} else {
			ZVAL_DEREF(op);
			if (Z_TYPE_P(op) != IS_STRING) {
				/* Wrong type, fall back to ZEND_CASE chain */
				ZEND_VM_NEXT_OPCODE();
			}
		}
	}

	jump_zv = zend_hash_find_ex(jumptable, Z_STR_P(op), OP1_TYPE == IS_CONST);
	if (jump_zv != NULL) {
		ZEND_VM_SET_RELATIVE_OPCODE(opline, Z_LVAL_P(jump_zv));
		ZEND_VM_CONTINUE();
	} else {
		/* default */
		ZEND_VM_SET_RELATIVE_OPCODE(opline, opline->extended_value);
		ZEND_VM_CONTINUE();
	}
}
```

There's two possible approaches:

1. Emit a brand new opcode type ZEND_MATCH_STRING and ZEND_MATCH_LONG, that would unconditionally jump to the `default` branch for (i.e. `ZEND_VM_SET_RELATIVE_OPCODE(opline, opline->extended_value)` - I assume `jump_zv != NULL` will always be true?)

   That'd be my preference, but others on internals might prefer to leave this as-is (e.g. if types would usually match in practice)
2. Move the default case first - This will probably break assumptions made by opcache or the JIT (ext/opcache/Optimizer has optimization passes for ZEND_SWITCH_LONG and ZEND_SWITCH_STRING)

Implementing opcache optimizations and the rest of the code that goes with it is time consuming and difficult in practice, especially for parts of opcache one's unfamiliar with, so I'm not recommending changing anything here, just adding a note on how this could be optimized in the future if it's deemed worth it.

Reply:
> all of the comparisons would be performed and fail before falling through to the default or throw UnhandledMatchError case.

This is probably way I'll do it: Generate all the arms with const conditions first (as the order for those doesn't matter) and generate a jump table for those. If the jump table fails move on two the non-constant conditions. When those fail it'll just move to the default case.

This should cover all the cases.

My test was actually wrong, `ZEND_SWITCH_STRING` and `ZEND_SWITCH_LONG` already do type checks. But yes we need to make sure there are no wrong assumptions in the optimizer.

Reply:
Hmm... unfortunately not quite as simple as I thought.

```php
```

Reply:
A potential optimization we could apply to both `switch` and `match` is the following:

```php
match ($x) {
    1 => ...,
    'a' => ...,
    2 => ...,
    'b' => ...,
    $y => ...,
    3 => ...,
    'c' => ...,
    4 => ...,
    'd' => ...,
}

// Can be optimized to:
0001: MATCH_LONG $x, 1 => 0002, 2 => 0003, default => 0004
0002: ...
0003: ...
0004: MATCH_LONG $x, 'a' => 0005, 'b' => 0006, default => 0007
0005: ...
0006: ...
0007: IS_IDENTICAL $x, $y
...
0010: MATCH_LONG $x, 3 => 0011, 4 => 0012, default => 0013
0011: ...
0012: ...
0013: MATCH_LONG $x, 'c' => 0014, 'd' => 0015, default => 0016
0014: ...
0015: ...
```

In other words, we can group constant conditions together and create multiple jumptables per match/switch. Not sure how common code like this is in the real world.

Reply:
I'll resolve this for now. Maybe we can improve the jumptable further later on.